library/ssl_msg.c

```diff
@@ -2560,9 +2560,15 @@ int mbedtls_ssl_write_record( mbedtls_ssl_context *ssl, uint8_t force_flush )
 #endif
         /* Skip writing the record content type to after the encryption,
          * as it may change when using the CID extension. */
-
-        mbedtls_ssl_write_version( ssl->major_ver, ssl->minor_ver,
-                           ssl->conf->transport, ssl->out_hdr + 1 );
+        int minor_ver = ssl->minor_ver;
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
+        /* TLS 1.3 still uses the TLS 1.2 version identifier
+         * for backwards compatibility. */
+        if( minor_ver == MBEDTLS_SSL_MINOR_VERSION_4 )
+            minor_ver = MBEDTLS_SSL_MINOR_VERSION_3;
+#endif /* MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL */
+        mbedtls_ssl_write_version( ssl->major_ver, minor_ver,
```

Comment:
As we support only TLS 1.2 and TLS 1.3, and as this field is deprecated by TLS 1.3 it could even be:
```suggestion
        mbedtls_ssl_write_version( MBEDTLS_SSL_MAJOR_VERSION_3, MBEDTLS_SSL_MINOR_VERSION_3,
```
with a comment explaining those "hardcoded" major and minor versions.

Reply:
Got CI build fail
```
Not supported version check: srv max TLS 1.0 ........................... FAIL

  ! pattern 'Error in protocol version' MUST be present in the Server output

  ! outputs saved to o-XXX-307.log

Not supported version check: srv max TLS 1.1 ........................... FAIL

  ! pattern 'Error in protocol version' MUST be present in the Server output

  ! outputs saved to o-XXX-308.log
```

Reply:
I just remove the change.

Reply:
I am a bit curious about that. Did you get those errors on the CI or when running ssl-opt.sh locally ?

Reply:
On my side I get those errors when running the tests against `development` with GnuTLS 3.7.2 but not with GnuTLS 3.5.18. Thus isn't it the case that those errors are just due to the fact that some output messages have been changed in GnuTLS more recent versions.

Reply:
I do not think so.  After I remove it , it pass the test. 

I will try dig root cause tomorrow.

Reply:
Fail CI link: https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-head/view/change-requests/job/PR-5120-head/5/artifact/

GnuTLS version is gnutls-3.4.10 . it reports packet errors
ssl_client2 reports `SSL - Handshake protocol not within min/max boundaries`

Reply:
I got strange result .
with same `gnutls-serv`, run `ssl-opt.sh` , No `Error in protocol version` .
Manual run command in the test, I got the message.

Reply:
I have got the root cause. We can not do hardcoded version .
If the server is TLS1.0/TLS1.1, client should reply a alert message with TLS1.0/TLS1.1 version HEAD. 
- Without hardcoded, the log is
```
ssl_msg.c:4716: |3| send alert level=2 message=70
ssl_msg.c:2550: |2| => write record
ssl_msg.c:2642: |3| output record: msgtype = 21, version = [3:2], msglen = 2
ssl_msg.c:2645: |4| dumping 'output record sent to network' (7 bytes)
ssl_msg.c:2645: |4| 0000:  15 03 02 00 02 02 46                             ......F
```
- With hardcoded
```
ssl_msg.c:4718: |3| send alert level=2 message=70
ssl_msg.c:2550: |2| => write record
ssl_msg.c:2644: |3| output record: msgtype = 21, version = [3:3], msglen = 2
ssl_msg.c:2647: |4| dumping 'output record sent to network' (7 bytes)
ssl_msg.c:2647: |4| 0000:  15 03 03 00 02 02 46                             ......F
```

Reply:
Another issue is about GnuTLS . I test 3.4.10 and 3.7.2 . 
For 3.7.2 , if stdout is redirect to a nomal file , below lines will not output.
```
* Accepted connection from IPv4 127.0.0.1 port 44470 on Thu Nov 11 03:15:50 2021
* Received alert '70': Error in protocol version.
```

Reply:
Thanks for the investigation. I understand now why it cannot be hard-coded.